File: 0501-1000/0680. Valid Palindrome II.go
package main
// Easy
// Time:O(n), Space:O(1)
// https://leetcode.cn/problems/valid-palindrome-ii/

func validPalindrome(s string) bool {
    left, right := 0, len(s) - 1
    
    for left < right {
        if s[left] != s[right] {
            // del left or right
            return isValid(s, left + 1, right) || isValid(s, left, right - 1)
        }
        left++
        right--
    }
    return true
}

func isValid(s string, left, right int) bool {
    for left < right {
        if s[left] != s[right] {
            return false
        }
        left++
        right--
    }
    return true
}

File: 0501-1000/0680. Valid Palindrome II.py
# Time:O(n), Space:O(1)

class Solution:
    def validPalindrome(self, s: str) -> bool:
        def is_valid(left, right):
            while left < right:
                if s[left] != s[right]:
                    return False
                left += 1
                right -= 1

            return True

        left, right = 0, len(s) - 1

        while left < right:
            if s[left] != s[right]:
                return is_valid(left + 1, right) or is_valid(left, right - 1)
            left += 1
            right -= 1

        return True
